### PR TITLE
Add the output basename to the lld.id symlink

### LINKED ISSUE #6979: `zig cc -o` broken?
```sh
zig cc -o app app.c
```

```text
error: FileNotFound
```

`zig` doesn't get that `-o` requires a parameter and  is trying to open the nonexistent `app` file as if it were source code:

```text
openat(7, "app", O_RDONLY|O_NOCTTY|O_LARGEFILE|O_CLOEXEC) = -1 ENOENT (No such file or directory)
```

### PR DESCRIPTION
**Pull request by alexnask**
linkWithLLD functions now use the symlink to also keep track of the current basename of the linking output and rename it as necessary.  

This closes #6979  
The way this is currently implemented, a maximum length of 223 bytes is imposed on the basename of the linking output when linking with LLD, in order to guarantee it will fit in the 255 bytes of symlink metadata.  

This limit can be removed by making lld.id a file instead of a symlink, however I believe it is reasonable.

### CODE REVIEW

File: src/link/Wasm.zig [hunk 1]
```diff
@@ -309,17 +316,44 @@ fn linkWithLLD(self: *Wasm, comp: *Compilation) !void {
...
+            prev_id_data
+        else
+            prev_id_data[0..digest.len];
+
+        if (mem.eql(u8, prev_digest, &digest)) blk: {
+            if (!mem.eql(u8, prev_id_data[digest.len..], sub_path)) {
+                // Rename the output to the new basename
+                directory.handle.rename(prev_id_data[digest.len..], sub_path) catch |err| {
+                    // If the rename fails, handle this as a cache miss.
+                    log.debug("Error renaming MachO linking output: {}", .{@errorName(err)});
```
Comment (xackus) [inline src/link/Wasm.zig:339]: There's got to be a better way than copying this code four times.
`MachO` -> `Wasm`

Comment (alexnask) [reply src/link/Wasm.zig:339]: Thanks, will fix in a minute.  
I agree, I want to unify those somehow in another PR.

**Comment by andrewrk** [comment]
I'd like to offer this alternative patch to solve the issue:

```diff
diff --git a/src/Cache.zig b/src/Cache.zig
index dd96e68d3..3295ffac7 100644
--- a/src/Cache.zig
+++ b/src/Cache.zig
@@ -60,6 +60,8 @@ pub const File = struct {
 pub const HashHelper = struct {
     hasher: Hasher = hasher_init,
 
+    const EmitLoc = @import("Compilation.zig").EmitLoc;
+
     /// Record a slice of bytes as an dependency of the process being cached
     pub fn addBytes(hh: *HashHelper, bytes: []const u8) void {
         hh.hasher.update(mem.asBytes(&bytes.len));
@@ -71,6 +73,15 @@ pub const HashHelper = struct {
         hh.addBytes(optional_bytes orelse return);
     }
 
+    pub fn addEmitLoc(hh: *HashHelper, emit_loc: EmitLoc) void {
+        hh.addBytes(emit_loc.basename);
+    }
+
+    pub fn addOptionalEmitLoc(hh: *HashHelper, optional_emit_loc: ?EmitLoc) void {
+        hh.add(optional_emit_loc != null);
+        hh.addEmitLoc(optional_emit_loc orelse return);
+    }
+
     pub fn addListOfBytes(hh: *HashHelper, list_of_bytes: []const []const u8) void {
         hh.add(list_of_bytes.len);
         for (list_of_bytes) |bytes| hh.addBytes(bytes);
diff --git a/src/Compilation.zig b/src/Compilation.zig
index 863cd35ba..91a191502 100644
--- a/src/Compilation.zig
+++ b/src/Compilation.zig
@@ -708,7 +708,7 @@ pub fn create(gpa: *Allocator, options: InitOptions) !*Compilation {
         cache.hash.add(options.link_libcpp);
         cache.hash.add(options.output_mode);
         cache.hash.add(options.machine_code_model);
-        cache.hash.add(options.emit_bin != null);
+        cache.hash.addOptionalEmitLoc(options.emit_bin);
         cache.hash.addBytes(options.root_name);
         // TODO audit this and make sure everything is in it
 
@@ -2786,11 +2786,11 @@ fn updateStage1Module(comp: *Compilation, main_progress_node: *std.Progress.Node
     man.hash.add(comp.bin_file.options.function_sections);
     man.hash.add(comp.bin_file.options.is_test);
     man.hash.add(comp.bin_file.options.emit != null);
-    man.hash.add(comp.emit_h != null);
-    man.hash.add(comp.emit_asm != null);
-    man.hash.add(comp.emit_llvm_ir != null);
-    man.hash.add(comp.emit_analysis != null);
-    man.hash.add(comp.emit_docs != null);
+    man.hash.addOptionalEmitLoc(comp.emit_h);
+    man.hash.addOptionalEmitLoc(comp.emit_asm);
+    man.hash.addOptionalEmitLoc(comp.emit_llvm_ir);
+    man.hash.addOptionalEmitLoc(comp.emit_analysis);
+    man.hash.addOptionalEmitLoc(comp.emit_docs);
     man.hash.addOptionalBytes(comp.test_filter);
     man.hash.addOptionalBytes(comp.test_name_prefix);
```

Pro: it's much simpler & less code to maintain
Pro: faster for common use cases
Cons: less optimal for the use case where someone wants to use `-femit-bin=foo` for one-time use without recompiling everything. (in the case where they regularly use the same `-femit-bin=foo` value it would be fine)

Are you OK with this, @alexnask?

**Comment by andrewrk** [comment]
In the interest of progress, I'm going to move forward with my counter-proposal, but @alexnask please feel free to open a proposal if you want to re-evaluate your other idea here.

**Pull request closed without merging**